pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -1390,125 +1224,67 @@ private class IndexContainer implements Closeable {
       _valueAggregator = valueAggregator;
     }
 
     DataSource toDataSource() {
       return new MutableDataSource(_fieldSpec, _numDocsIndexed, _valuesInfo._numValues,
           _valuesInfo._maxNumValuesPerMVEntry, _dictionary == null ? -1 : _dictionary.length(), _partitionFunction,
-          _partitions, _minValue, _maxValue, _forwardIndex, _dictionary, _invertedIndex, _rangeIndex, _textIndex,
-          _fstIndex, _jsonIndex, _h3Index, _bloomFilter, _nullValueVector, _valuesInfo._varByteMVMaxRowLengthInBytes);
+          _partitions, _minValue, _maxValue, _mutableIndexes, _dictionary, _nullValueVector,
+          _valuesInfo._varByteMVMaxRowLengthInBytes);
     }
 
     @Override
     public void close() {
       String column = _fieldSpec.getName();
-      try {
-        _forwardIndex.close();
-      } catch (Exception e) {
-        _logger.error("Caught exception while closing forward index for column: {}, continuing with error", column, e);
-      }
-      if (_dictionary != null) {
-        try {
-          _dictionary.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing dictionary for column: {}, continuing with error", column, e);
-        }
-      }
-      if (_invertedIndex != null) {
-        try {
-          _invertedIndex.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing inverted index for column: {}, continuing with error", column,
-              e);
-        }
-      }
-      if (_rangeIndex != null) {
-        try {
-          _rangeIndex.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing range index for column: {}, continuing with error", column, e);
-        }
-      }
-      if (_textIndex != null) {
-        try {
-          _textIndex.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing text index for column: {}, continuing with error", column, e);
-        }
-      }
-      if (_fstIndex != null) {
-        try {
-          _fstIndex.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing fst index for column: {}, continuing with error", column, e);
-        }
-      }
-      if (_jsonIndex != null) {
-        try {
-          _jsonIndex.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing json index for column: {}, continuing with error", column, e);
-        }
-      }
-      if (_h3Index != null) {
-        try {
-          _h3Index.close();
-        } catch (Exception e) {
-          _logger.error("Caught exception while closing H3 index for column: {}, continuing with error", column, e);
-        }
-      }
-      if (_bloomFilter != null) {
+
+      BiConsumer<IndexType<?, ?, ?>, AutoCloseable> closer = (indexType, closeable) -> {
         try {
-          _bloomFilter.close();
+          if (closeable != null) {
```

Comment:
Why would an index be null? Is this specifically for _dictionary and _nullValueVector?